app/controllers/users_controller.rb

```diff
@@ -76,39 +118,37 @@ def new
                                    :auth_uid => params[:auth_uid])
 
       flash.now[:notice] = render_to_string :partial => "auth_association"
+
+      # validate before displaying to show email field if it conflicts
+      current_user.valid?
     else
       check_signup_allowed
 
       self.current_user = User.new
+      redirect_to :action => :new
     end
   end
 
-  def create
+  def create_association
     self.current_user = User.new(user_params)
 
     if check_signup_allowed(current_user.email)
       session[:referer] = safe_referer(params[:referer]) if params[:referer]
 
-      Rails.logger.info "create: #{session[:referer]}"
+      Rails.logger.info "create_association: #{session[:referer]}"
 
-      if current_user.auth_provider.present? && current_user.pass_crypt.empty?
-        # We are creating an account with external authentication and
-        # no password was specified so create a random one
-        current_user.pass_crypt = SecureRandom.base64(16)
-        current_user.pass_crypt_confirmation = current_user.pass_crypt
-      end
+      # We are creating an account with external authentication and
+      # no password was specified so create a random one
+      current_user.pass_crypt = SecureRandom.base64(16)
+      current_user.pass_crypt_confirmation = current_user.pass_crypt
 
-      if current_user.invalid?
+      if current_user.invalid? || current_user.auth_provider.blank?
         # Something is wrong with a new user, so rerender the form
-        render :action => "new"
-      elsif current_user.auth_provider.present?
+        render :action => "new_association"
+      else
         # Verify external authenticator before moving on
         session[:new_user] = current_user.attributes.slice("email", "display_name", "pass_crypt", "consider_pd")
         redirect_to auth_url(current_user.auth_provider, current_user.auth_uid), :status => :temporary_redirect
```

Comment:
I have re-arranged the code to avoid round-trip to the provider, and redirect to new-association instead. I also removed the use of `session[:new_user]`, since all the info from 3rd party provider is already present when rendering the page. In order to prevent users from tempering the email address received from provider, I pass around sha256 hmac instead of storing it in `session[:new_user]["verified_email"]`.